### PR TITLE
Fix bool array warping

### LINKED ISSUE #4292: Rescaling boolean arrays yields plausible result or array of zeros depending on the scale factor
## Description
Hello!
I am using `skimage.transform.rescale` to down-scale boolean arrays that have a single connected component.
Depending on the scale factor selected, the output yields a result that seems correct or an array full of zeros.

If I convert the boolean array to float, the result is different than when I pass the array with values as bool or int. 
Setting 'anti-aliasing' to False yields the same result for int, bool or float values in the array.

Furthermore, the expression given in the `rescale` function documentation to compute the "anti_aliasing_sigma" is different than the actual expression used inside `skimage.transform.resize` (which is called in `skimage.transform.rescale`):
- `skimage.transform.rescale`: "By default, this value is chosen as (1 - s) / 2 where s is the down-scaling factor."
- `skimage.transform.resize`: "By default, this value is chosen as (s - 1) / 2 where s is the down-scaling factor, where s > 1."

## Way to reproduce
```python
import numpy as np
from skimage.transform import rescale

input_matrix_shape = (1000, 1200, 400)
high_resolution_matrix = np.zeros(input_matrix_shape)
high_resolution_matrix[200:800, 300:900, 100:300] = 1 # Fill matrix with some ones
high_resolution_matrix = high_resolution_matrix.astype(np.bool_)
scale_factor_list = [0.625, 0.305]

for current_scale_factor in scale_factor_list:
    rescaled_matrix = rescale(high_resolution_matrix, current_scale_factor, order = 0, anti_aliasing=True, multichannel=False, preserve_range=True)
    print('Scale factor: {}'.format(current_scale_factor))
    print('\tResult shape: {}'.format(rescaled_matrix.shape))
    print('\tTotal non-zero values: {}'.format(np.sum(rescaled_matrix > 0)))
    print()
```

## Version information
```python
# Paste the output of the following python commands
from __future__ import print_function
import sys; print(sys.version)
import platform; print(platform.platform())
import skimage; print("scikit-image version: {}".format(skimage.__version__))
import numpy; print("numpy version: {}".format(numpy.__version__))
```

```python
3.7.5 (default, Oct 17 2019, 12:21:00) 
[GCC 8.3.1 20190223 (Red Hat 8.3.1-2)]
Linux-5.2.17-100.fc29.x86_64-x86_64-with-fedora-29-Twenty_Nine
scikit-image version: 0.16.2
numpy version: 1.17.3
```

### PR DESCRIPTION
**Pull request by rfezzani**
## Description

Fixes #4292.

This PR proposes
- to set default spline interpolation's order to 0 if input array dtype is bool and 1 otherwise. The function `skimage._shared.utils._set_order` is created for this purpose and is used wherever the `order` argument is used (`skimage.transform._warps` module and `skimage.measure.profile.profile_line` function).
- to set default anti_aliasing parameter to False if input array dtype is bool and True otherwise.

~~No deprecation cycle is started since it is a bug fix, do you think it should?~~

## Checklist

- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [x] Unit tests
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

**Comment by rfezzani** [comment]
Gentle up for this PR :slightly_smiling_face:

File: skimage/_shared/utils.py [hunk 1]
```diff
@@ -286,3 +286,41 @@ def convert_to_float(image, preserve_range):
     else:
         image = img_as_float(image)
     return image
+
+
+def _set_order(image_dtype, order):
```
Comment (emmanuelle) [inline skimage/_shared/utils.py:291]: `_set_interpolation_order` would be more informative

Comment (emmanuelle) [reply skimage/_shared/utils.py:291]: `_validate_interpolation_order` or `_coerce_interpolation_order` would be even better.

File: skimage/transform/_warps.py [hunk 2]
```diff
@@ -60,9 +61,10 @@ def resize(image, output_shape, order=1, mode='reflect', cval=0, clip=True,
         image is converted according to the conventions of `img_as_float`.
         Also see https://scikit-image.org/docs/dev/user_guide/data_types.html
     anti_aliasing : bool, optional
-        Whether to apply a Gaussian filter to smooth the image prior to
-        down-scaling. It is crucial to filter when down-sampling the image to
-        avoid aliasing artifacts.
+        Whether to apply a Gaussian filter to smooth the image prior
+        to down-scaling. It is crucial to filter when down-sampling
+        the image to avoid aliasing artifacts. If input image data
+        type is bool, no anti-aliasing is applyed.
```
Comment (emmanuelle) [inline skimage/transform/_warps.py:67]:
```suggestion
        type is bool, no anti-aliasing is applied.
```

File: skimage/transform/_warps.py [hunk 3]
```diff
@@ -231,9 +246,10 @@ def rescale(image, scale, order=1, mode='reflect', cval=0, clip=True,
         Whether the last axis of the image is to be interpreted as multiple
         channels or another spatial dimension.
     anti_aliasing : bool, optional
-        Whether to apply a Gaussian filter to smooth the image prior to
-        down-scaling. It is crucial to filter when down-sampling the image to
-        avoid aliasing artifacts.
+        Whether to apply a Gaussian filter to smooth the image prior
+        to down-scaling. It is crucial to filter when down-sampling
+        the image to avoid aliasing artifacts. If input image data
+        type is bool, no anti-aliasing is applyed.
```
Comment (emmanuelle) [inline skimage/transform/_warps.py:252]:
```suggestion
        type is bool, no anti-aliasing is applied.
```

**Comment by emmanuelle** [comment]
Thanks for the PR @rfezzani. So we're doing something different for bool and uint8 arrays with this change (was it already the case?). I don't mind this but should this be said in an even louder way in the docstrings?

**Comment by rfezzani** [comment]
Thank you @emmanuelle for your review!
The bugs and issues related to bool array transformation comes from the fact that interpolation and linear combination is not defined for boolean values. So this PR simply proposes to avoid anti-aliasing and interpolation with `bool` arrays.
The behavior of `uint8` arrays doesn't change here (linear interpolation and anti-aliasing by default)

**Comment by rfezzani** [comment]
:tada: Thank you @soupault!

**Comment by emmanuelle** [comment]
Thank you @rfezzani merging!

**Comment by hmaarrfk** [comment]
Thanks!

I think thisis a really nice way to address this issue!

**Pull request merged**